### ISSUE TITLE
GM injection in lumbar T2w imaging

### ISSUE THREAD

**Issue by Nilser3** [ivadomed/model-gm-contrast-region-agnostic#5]
# Description 

The GM image at the lumbo-sacral level looks very different compared to the cervical GM, so when training a region-agnostic GM segmentation algorithm, more lumbar image data is required. 

![image](https://github.com/user-attachments/assets/f6d0ef85-cb29-4552-bcbb-bbfa7e22834b)

We only have one dataset with visible lumbar GM (`lumbar-vanderbilt`) on images in T2S sequence.

## Proposition

There is a data published in this paper: [An open-access lumbosacral spine MRI dataset with enhanced spinal nerve root structure resolution](https://www.nature.com/articles/s41597-024-03919-4), it contains T2w sagittal whole spine and axial T2 images of the lumbo-sacral spinal cord (GM not visible):

![image](https://github.com/user-attachments/assets/792bb97f-6c05-4f95-bf74-25709088e4e2)

The idea is to register the axial T2w images to the PAM50 template, in order to obtain a probabilistic GM mask in the subject space.

For this, we can use the `seg_sc_contrast_agnostic` model to segment the SC in the axial T2w image and we can extract the intervertebral discs labeling by applying the `totalspineseg` model in the sagittal T2w image.

With which we could apply: 

```
sct_register_to_template  -i "$input_ax-T2w" -s "$input_ax-T2w_SC" -ldisc  "$input_sag-T2w_labels"
```
and by extracting the probabilistic masks of the GM of the PAM50 we would obtain: 

![image](https://github.com/user-attachments/assets/544679c8-8135-4205-b083-f596d6866f00)

Multiplying the soft mask by the image_T2w_ax, and add this product to the image image_T2w_ax we would inject a GM image, also by binarizing the soft mask, we would obtain a GM segmentation (see [my code](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/synth_gm_data.py)).

### Preliminary results:

![image](https://github.com/user-attachments/assets/5236e252-ef0e-48e3-b38c-59619eae7784)

![image](https://github.com/user-attachments/assets/4e47f5d8-d28a-48be-ab99-8a0e65dc527c)

## Related issues: 
#2

**Comment by Nilser3** [comment]
## Preliminary results on `lumbar-marseille` dataset:

The difference in this case is that both the `seg_sc_contrast_agnostic` and `totalspineseg` models were applied to the ax-T2w image (good results labeling the intervertebral discs).

**sub-CTS09**
![image](https://github.com/user-attachments/assets/4bff7e19-7650-457f-8e72-262fd7f92cff)

![image](https://github.com/user-attachments/assets/88eb1eb4-5a26-4f45-ae13-b2b0db2cafa1)

## Next steps:

- [x] Train a model GM lumbar-seg based on real contrast (`lumbar-vanderbild`)
- [x] Train a model GM lumbar-seg based on real + synthetic contrast 
- [x] Compare both models 

Feedback pls @jcohenadad

**Comment by jcohenadad** [comment]
@Nilser3 This is a very interesting and original approach. At first, I would go against this 'fake' creation of GM masks, but if you can prove that it provides better model performance, then I would say go for it!

**Comment by Nilser3** [comment]
As a proof of concept I trained 2 models:

1. Model real data: (`lumbar-vanderbild`) 
- N of 2D images : 561 for train/valid

2. Model real + synth data: (`lumbar-vanderbild` + `shanghai-synth` + `lumbar-marseille-synth`)
- N of 2D images: 561 + 483 + 960 = 2004 for train/valid

Both are 2D nnUnetV2 models, same fold 0, and same real data for `testing`: 12 subjects (166 2D images )from `lumbar-vanderbild`.

# Preliminary results:

Comparison of Dice Score measured on 2D slices (Stat using Wilcoxon test)

<img src="https://github.com/user-attachments/assets/b55d2316-26da-4678-8310-193d6dd471ab" alt="Image" width="400"/>


### Stat description:

| Statistic   | Model real data | Model real + synth data|
|-------------|-------------|-------------------|
| Count       | 166.000000  | 166.000000        |
| Mean        | 0.883873    | 0.888816          |
| Std         | 0.057540    | 0.053268          |
| Min         | 0.620155    | 0.588235          |
| 25%         | 0.862014    | 0.871718          |
| 50%         | 0.900974    | 0.899782          |
| 75%         | 0.919583    | 0.920898          |
| Max         | 0.969231    | 0.963563          |

### Qualitative results  on `sub-242174`
Color legend : 
- red -> manual GT
- white -> Model real data
- green -> Model real + synth data

<img src="https://github.com/user-attachments/assets/6edb246f-329a-4bf5-a430-47d3341439e0" alt="Image" width="300"/>

<img src="https://github.com/user-attachments/assets/f1ca2326-869d-4811-8dd5-89f85a7cac0c" alt="Image" width="300"/>

<img src="https://github.com/user-attachments/assets/3ed5787b-4d3d-44b1-b650-75eb683c685f" alt="Image" width="300"/>

<img src="https://github.com/user-attachments/assets/123b8a01-2ac9-40b4-b182-64f95b070e1f" alt="Image" width="300"/>

but in the same subject, we have some slices where both models have failures:

<img src="https://github.com/user-attachments/assets/39d9aa57-8d7e-40d8-9f92-51a56a39cf64" alt="Image" width="300"/>
<img src="https://github.com/user-attachments/assets/b487757e-b9f6-4e66-ba55-fef7d012abc4" alt="Image" width="300"/>

Although in these slices start GTs are weird, 

I think there is a contribution of the synth-data in the lumbar GM segmentation. 

Maybe we should add it to the contrast-region-agnostic training database.

**Comment by Nilser3** [comment]
Both models will be tested in a new dataset coming soon: `basel-rAMIRA` 

![Image](https://github.com/user-attachments/assets/0fcd0216-263a-4c02-adc8-2cd8f881d773)
From: [Cervical and thoracic spinal cord gray matter atrophy isassociated with disability in patients with amyotrophic lateralsclerosis](https://onlinelibrary.wiley.com/doi/10.1111/ene.16268)

**Comment by vcallot** [comment]
When using the template, you will always add the "same" information (about the GM shape), so there will be an over redundancy of the information, without creating/using variability. This is probably a strong bias, to be taken into account.

**Comment by Nilser3** [comment]
to test on a external dataset: `marseille-t2s-template`

**Comment by jcohenadad** [comment]
about https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/5#issuecomment-2548030738 

> Both are 2D nnUnetV2 models, same fold 0, and same real data for testing: 12 subjects (166 2D images) from lumbar-vanderbild.

Important to also test on out-of-distribution. Two possibilities for now:
- PAM50 lumbar T2*
- train only with synthetic data and test on vanderbilt

**Comment by Nilser3** [comment]
# Model Fully Trained on Synthetically Generated Data  

This is a model trained exclusively with synthetic data, using the following datasets:  

- **marseille-lumbar**: T2w, 10 subjects / 1180 2D [denoised](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/run_denoising.sh) and [ghosted](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/run_ghost.sh) images 
- **shanghai-lumbar**: T2w, 14 subjects / 617 2D [denoised](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/run_denoising.sh) and [ghosted](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/run_ghost.sh) images
- **bavaria-quebec-spine-ms-unstitched (chunk inferior)**: T2w, 65 subjects / 1283 2D [ghosted](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/run_ghost.sh) images  

After applying [synth_gm_data.py](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/synth_gm_data.py) script, two models were trained on 250 epochs:

<details><summary>2D</summary>

```console
"2d": {
            "data_identifier": "nnUNetPlans_2d",
            "preprocessor_name": "DefaultPreprocessor",
            "batch_size": 46,
            "patch_size": [
                448,
                448
            ],
            "median_image_size_in_voxels": [
                410.0,
                410.0
            ],
            "spacing": [
                0.29296875,
                0.29296875
            ],
```

</details>

<details><summary>3D-fullres</summary>

```console
"3d_fullres": {
            "data_identifier": "nnUNetPlans_3d_fullres",
            "preprocessor_name": "DefaultPreprocessor",
            "batch_size": 2,
            "patch_size": [
                160,
                224,
                192
            ],
            "median_image_size_in_voxels": [
                330.0,
                410.0,
                410.0
            ],
            "spacing": [
                0.5000009536743164,
                0.29296875,
                0.29296875
            ],

```
</details>


showing promising results.

## Results  

1. **lumbar-vanderbilt**:  
   53 subjects / 727 2D images  
<img src="https://github.com/user-attachments/assets/d121bf10-5979-47e5-81be-0190370d208c" alt="Image" width="400"/>

2. **marseille-t2s-template**:  
   25 subjects / 345 2D images  
![Image](https://github.com/user-attachments/assets/d65e39fb-f3a3-47de-922e-50f80a42b9aa)

3. **PAM50_t2s.nii.gz image**:  
   1 volume / 828 2D images  
![Image](https://github.com/user-attachments/assets/63c92048-4061-49c3-b93a-b0916fbe7aac)

I find it interesting to see that the `2D` model performs better at lower levels, and that the `3D_fullres` model segment better at higher levels because it generalizes with the 3D context information.

Following the philosophy of the arXiv [Can segmentation models be trained with fully synthetically generated data?](https://arxiv.org/pdf/2209.08256), I consider that we can add this synthetically generated data in the training of a contrast-region-agnostic GM segmentation model.

**Comment by jcohenadad** [comment]
Some feedback:
- would be useful to see some augmented data (raw, denoised, ghosted) to have an idea how realistic that is
- consider varying the amount of ghosting (if not already done)-- is it done along AP? RL? both? in general ppl acquire data with phase encoding AP (hence with ghosting along AP)

**Comment by Nilser3** [comment]
After exploring and improving the PAM50 registrations to subject space and generating the synthetic images, it was observed that the generation of realistic lumbar butterflies (GM) is complicated, see image:

<img src="https://github.com/user-attachments/assets/872c2df3-a95f-4583-bdd4-ff22a9ec637f" alt="Image" width="500"/>

Therefore, inferences on real lumbar images of the synthetic data based models were always under-segmented, see image:

<img src="https://github.com/user-attachments/assets/7b981414-b997-4876-8cc7-d0a35eeadc47" alt="Image" width="300"/>

so it was decided not to merge this synthesized data into the `contrast-region-agnostic-gm` training database.
